lrmd/lrmd.c

```diff
@@ -160,6 +164,96 @@ build_rsc_from_xml(xmlNode * msg)
     return rsc;
 }
 
+static void
+dup_attr(gpointer key, gpointer value, gpointer user_data)
+{
+    g_hash_table_replace(user_data, strdup(key), strdup(value));
+}
+
+static gboolean
+valid_version_format(const char *version)
+{
+    if (version == NULL) {
+        return FALSE;
+    }
+
+    if (version_format_regex == NULL) {
+        const char *regex_string = "^[[:digit:]]+([.][[:digit:]]+)*$";
+        version_format_regex = calloc(1, sizeof(regex_t));
+        regcomp(version_format_regex, regex_string, REG_EXTENDED);
+    }
+
+    return regexec(version_format_regex, version, 0, NULL, 0) == 0;
+}
+
+static const char*
+get_ra_version(const char *class, const char *provider, const char *type, gboolean update_hash)
+{
+    int len = 0;
+    char *key = NULL;
+    const char *default_version = "0.1";
```

Comment:
Since this isn't static, the pointer is just pushed on the stack and isn't valid once the function returns. You could make this static, or return a strdup() of it, or use a #define constant instead (which would be a string literal, so it would stay in memory).